File: Fashion MNIST Classification.py
# Import ttthe necessary packages
import numpy as np
import matplotlib.pyplot as plt
from keras.datasets import fashion_mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Dense, Flatten
from tensorflow.keras.utils import to_categorical

# Load the Fashion MNIST dataset
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()
# Reshape the data to include a single channel (grayscale)
X_train = X_train.reshape((X_train.shape[0], 28, 28, 1))
X_test = X_test.reshape((X_test.shape[0], 28, 28, 1))

# scale data to the range of [0, 1]
X_train = X_train.astype('float32') / 255
X_test = X_test.astype('float32') / 255
# One-hot encode the training and testing labels
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

#Build the CNN model
model = Sequential()
 # Convolutional layer
    # filters with a kernel size of 5x5
model.add(Conv2D(64, (5, 5),
                      padding="same",
                      activation="relu", 
                      input_shape=(28, 28, 1)))

# MaxPooling layer
# Size with a kernel size of 2x2
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Conv2D(128, (5, 5), padding="same",
                      activation="relu"))


model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Conv2D(256, (5, 5), padding="same", 
                      activation="relu"))

     
model.add(MaxPooling2D(pool_size=(2, 2)))

 # Once the convolutional and pooling 
    # operations are done the layer
    # is flattened and fully connected layers
    # are added
model.add(Flatten())
model.add(Dense(256, activation="relu"))

# Output Layer: Dense layer with 10 neurons (one for each class)
model.add(Dense(10, activation='softmax'))

#Compile the method
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

#Train the method
history = model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=128)

# Evaluate model performance
test_loss, test_accuracy = model.evaluate(X_test, y_test)
print(f"Test accuracy: {test_accuracy:.4f}")

#2 Prediction
# Choose two sample images from the test set
sample_images = X_test[:2]
sample_labels = y_test[:2]

# Make predictions
predictions = model.predict(sample_images)
predicted_classes = np.argmax(predictions, axis=1)
actual_classes = np.argmax(sample_labels, axis=1)

# Display the results
for i in range(2):
    plt.imshow(sample_images[i].reshape(28, 28), cmap='gray')
    plt.title(f"Predicted: {predicted_classes[i]}, Actual: {actual_classes[i]}")
    plt.show()

File: fashion_Mnist classification.R

library(keras)
library(tensorflow)

# Load Fashion MNIST dataset
fashion_mnist <- dataset_fashion_mnist()
c(train_images, train_labels) %<-% fashion_mnist$train
c(test_images, test_labels) %<-% fashion_mnist$test

# Reshape the data for CNN input and normalize
train_images <- array_reshape(train_images, c(nrow(train_images), 28, 28, 1))
test_images <- array_reshape(test_images, c(nrow(test_images), 28, 28, 1))
train_images <- train_images / 255
test_images <- test_images / 255

# One-hot encode the labels
train_labels <- to_categorical(train_labels, 10)
test_labels <- to_categorical(test_labels, 10)

# Build the CNN Model
model <- keras_model_sequential() %>%
  layer_conv_2d(filters = 32, kernel_size = c(3, 3), activation = 'relu', input_shape = c(28, 28, 1)) %>%
  layer_max_pooling_2d(pool_size = c(2, 2)) %>%
  layer_conv_2d(filters = 64, kernel_size = c(3, 3), activation = 'relu') %>%
  layer_max_pooling_2d(pool_size = c(2, 2)) %>%
  layer_flatten() %>%
  layer_dense(units = 128, activation = 'relu') %>%
  layer_dense(units = 10, activation = 'softmax')

# Compile the model
model %>% compile(
  optimizer = 'adam',
  loss = 'categorical_crossentropy',
  metrics = c('accuracy')
)

# Train the model
history <- model %>% fit(
  train_images, train_labels,
  epochs = 10,
  batch_size = 128,
  validation_split = 0.2
)

# Evaluate the model
score <- model %>% evaluate(test_images, test_labels)
cat('Test accuracy:', score$accuracy, "\n")

# Make predictions on two sample images
predictions <- model %>% predict(test_images[1:2, , , drop = FALSE])
predicted_classes <- apply(predictions, 1, which.max) - 1
actual_classes <- apply(test_labels[1:2, , drop = FALSE], 1, which.max) - 1

# Display the results
for (i in 1:2) {
  image(matrix(test_images[i, , , 1], nrow = 28, ncol = 28)[, 28:1], col = gray.colors(255))
  title(main = paste("Predicted:", predicted_classes[i], "Actual:", actual_classes[i]))
}
